File: src/strategy/impl/Breakout.ts
import type { Strategy, Signal, StrategyContext } from '../Strategy';
import type { MarketSnapshot } from '../../market/MarketDataService';

export const Breakout: Strategy = {
  name: 'breakout',
  suitableRegimes: ['ranging', 'high_volatility'],
  evaluate(snap: MarketSnapshot, _ctx: StrategyContext): Signal {
    const { atr14, atrPct, adxValue, rsi14 } = snap.indicators;
    const { close, high, low } = snap.ohlcv;
    const price = snap.lastPrice;

    if (close.length < 20) {
      return { action: 'hold', symbol: snap.symbol, strategy: 'breakout', confidence: 0, rationale: 'Not enough data' };
    }

    const lookback = 20;
    const recentHighs = high.slice(-lookback);
    const recentLows = low.slice(-lookback);
    const rangeHigh = Math.max(...recentHighs.slice(0, -1));
    const rangeLow = Math.min(...recentLows.slice(0, -1));

    const breakoutUp = price > rangeHigh && adxValue < 30 && rsi14 > 55;
    const breakoutDown = price < rangeLow && adxValue < 30 && rsi14 < 45;

    if (breakoutUp) {
      return {
        action: 'enter_long', symbol: snap.symbol, strategy: 'breakout',
        confidence: 0.65,
        suggestedEntry: price,
        suggestedStop: rangeLow,
        suggestedTp: price + (price - rangeLow) * 1.5,
        rationale: `Breakout above ${rangeHigh.toFixed(2)}, ATRpct=${(atrPct * 100).toFixed(2)}%`,
      };
    }
    if (breakoutDown) {
      return {
        action: 'enter_short', symbol: snap.symbol, strategy: 'breakout',
        confidence: 0.65,
        suggestedEntry: price,
        suggestedStop: rangeHigh,
        suggestedTp: price - (rangeHigh - price) * 1.5,
        rationale: `Breakdown below ${rangeLow.toFixed(2)}, ATR14=${atr14.toFixed(2)}`,
      };
    }

    return { action: 'hold', symbol: snap.symbol, strategy: 'breakout', confidence: 0, rationale: 'No breakout' };
  },
};

File: bybit_agent/strategy/impl/breakout.py
"""Breakout — ports src/strategy/impl/Breakout.ts."""

from __future__ import annotations

from ..base import Signal, StrategyContext


class Breakout:
    name = "breakout"
    suitable_regimes = ["ranging", "high_volatility"]

    def evaluate(self, snap, ctx: StrategyContext) -> Signal:
        i = snap.indicators
        close, high, low = snap.ohlcv["close"], snap.ohlcv["high"], snap.ohlcv["low"]
        price = snap.lastPrice

        if len(close) < 20:
            return Signal(action="hold", symbol=snap.symbol, strategy="breakout",
                          confidence=0, rationale="Not enough data")

        lookback = 20
        recent_highs = high[-lookback:]
        recent_lows = low[-lookback:]
        range_high = max(recent_highs[:-1])
        range_low = min(recent_lows[:-1])

        breakout_up = price > range_high and i.adxValue < 30 and i.rsi14 > 55
        breakout_down = price < range_low and i.adxValue < 30 and i.rsi14 < 45

        if breakout_up:
            return Signal(
                action="enter_long", symbol=snap.symbol, strategy="breakout", confidence=0.65,
                suggestedEntry=price, suggestedStop=range_low,
                suggestedTp=price + (price - range_low) * 1.5,
                rationale=f"Breakout above {range_high:.2f}, ATRpct={i.atrPct * 100:.2f}%",
            )
        if breakout_down:
            return Signal(
                action="enter_short", symbol=snap.symbol, strategy="breakout", confidence=0.65,
                suggestedEntry=price, suggestedStop=range_high,
                suggestedTp=price - (range_high - price) * 1.5,
                rationale=f"Breakdown below {range_low:.2f}, ATR14={i.atr14:.2f}",
            )
        return Signal(action="hold", symbol=snap.symbol, strategy="breakout",
                      confidence=0, rationale="No breakout")
